godot-core/src/engine/script_instance.rs

```diff
@@ -292,10 +225,13 @@ pub fn create_script_instance<T: ScriptInstance>(rs_instance: T) -> *mut c_void
     let instance_ptr = Box::into_raw(Box::new(gd_instance));
 
     let data = ScriptInstanceData {
-        inner: RefCell::new(rs_instance),
+        inner: GdCell::new(rs_instance),
         script_instance_ptr: instance_ptr,
         property_list: Default::default(),
         method_list: Default::default(),
+        // SAFETY: The script instance is always freed before the base object is destroyed. The weak reference should therefore never be
+        // accessed after it has been freed.
+        base: unsafe { Base::from_gd(&for_object) },
```

Comment:
What happens if the person calling `create_script_instance` also frees this object? like
```rs
let ptr = create_script_instance(.., object);
object.free();
ptr
```

Reply:
`create_script_instance` returns an opaque `c_void` raw pointer. If the object is freed before the pointer is passed back to Godot, it can never be used by Godot (I think it would actually crash the engine). If the weak-reference is cast into a `*mut __GdextScriptInstance` in rust and then accessed, the weak-reference could in theory be accessed as well.

Reply:
Can you check that it just crashes the engine or some other "graceful" thing, and doesn't cause UB by for instance seg faulting or something? i doubt we'd be able to actually write a test for this but still it'd be good to know for sure.

Reply:
Just tested this and its segfaults in this line of the engine code https://github.com/godotengine/godot/blob/a754bbc565ff39b8894951ebdd62ec1042b62ccc/core/object/object.cpp#L969. The engine tries to assign the script_instance field of the object, but the object has already been freed.

Reply:
ok so then this current api is unsound. one way to fix that could be to make `create_script_instance` an unsafe function with the safety invariant that `for_object` must not be freed for as long as the script instance is live/may be accessed by godot/something like that.

since causing a segfault is unacceptable for safe rust.

this should ideally be fixed by changing the `instance_create` virtual method but that's a much bigger change and not something that needs to be done here.

Reply:
I can make `create_script_instance` unsafe if necessary, but to clarify:

The segfault is caused by freeing `for_object` inside `unsafe fn instance_create(&self, for_object: Gd<Object>) -> *mut c_void` which is already `unsafe`.

Reply:
yes, `instance_create` is unsafe. which means the caller of the method must ensure that they dont call it in the wrong way. but the implementer is free to do whatever they want in the method body as long as they dont violate any additional safety invariants. for instance

```rs
unsafe fn foo() {
  transmute::<i32, bool>(10);
}
```

this function is unsound, because even though the function is `unsafe`, the caller cannot soundly call this method in any way. The only way this could be made sound would be if the safety doc on the function was something like
```rs
/// # Safety
/// Calling this function is UB.
```

Since then the caller is responsible for not calling the function.

Similarly, the documentation of the `instance_create` method could be changed to say something like
```rs
/// # Safety
/// The caller must ensure that the implementation of this method doesn't free `for_object`
```
However there are several issues here:
1. You should not need to inspect a method's implementation to know if it's safe to call it or not.
2. The person implementing `instance_create` can just choose to overwrite this safety documentation.
3. This makes it impossible to generically call the method since you cannot know the implementation beforehand. (which we effectively do in gdext)

Rust has a way to deal with this, and it's `unsafe` traits. However that would require a bigger codegen change. So for now i think the easiest way to at least avoid this unsoundness issue would be to make `instance_create` an `unsafe` function so that this specific case would have an actual `unsafe` block that would cause the UB.

Reply:
Thanks for the thorough explanation. This makes a lot of sense.